File: test/core/utils/contact_email_test.dart
import 'package:career_gamification/core/utils/contact_email.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('ContactEmail', () {
    test('normaliza e valida um contato comum', () {
      expect(
        ContactEmail.normalize('  Pessoa@Example.COM '),
        'pessoa@example.com',
      );
      expect(ContactEmail.isUsable('Pessoa@Example.COM'), isTrue);
    });

    test('detecta os dois domínios exatos do Apple Private Relay', () {
      expect(
        ContactEmail.isApplePrivateRelay('abc@privaterelay.appleid.com'),
        isTrue,
      );
      expect(
        ContactEmail.isApplePrivateRelay('abc@private.icloud.com'),
        isTrue,
      );
      expect(
        ContactEmail.isApplePrivateRelay('abc@sub.private.icloud.com'),
        isFalse,
      );
      expect(ContactEmail.isUsable('abc@private.icloud.com'), isFalse);
    });

    test('não confunde um e-mail pessoal do iCloud com Private Relay', () {
      expect(ContactEmail.isApplePrivateRelay('pessoa@icloud.com'), isFalse);
      expect(ContactEmail.isUsable('pessoa@icloud.com'), isTrue);
    });

    test('detecta endereço sintético do login por telefone', () {
      expect(
        ContactEmail.isSyntheticAuthEmail('phone_5511987654321@stage.app'),
        isTrue,
      );
      expect(ContactEmail.isUsable('phone_5511987654321@stage.app'), isFalse);
      expect(ContactEmail.isSyntheticAuthEmail('pessoa@stage.app'), isFalse);
    });

    test('rejeita formatos inválidos', () {
      expect(ContactEmail.isUsable('sem-arroba.example.com'), isFalse);
      expect(ContactEmail.isUsable('a@@example.com'), isFalse);
      expect(ContactEmail.resumeValueOrEmpty('invalido'), isEmpty);
    });

    test('resolve perfil, extração e autenticação nessa ordem', () {
      expect(
        ContactEmail.resolveInitial(
          profileEmail: 'perfil@example.com',
          extractedEmail: 'cv@example.com',
          authEmail: 'login@example.com',
        ),
        'perfil@example.com',
      );

      expect(
        ContactEmail.resolveInitial(
          profileEmail: 'relay@privaterelay.appleid.com',
          extractedEmail: 'CV@Example.com',
          authEmail: 'login@example.com',
        ),
        'cv@example.com',
      );

      expect(
        ContactEmail.resolveInitial(
          profileEmail: 'phone_5511999999999@stage.app',
          extractedEmail: 'invalido',
          authEmail: 'relay@private.icloud.com',
        ),
        isEmpty,
      );
    });
  });
}

File: supabase/functions/_shared/contact_email.test.ts
import { assertEquals, assertFalse } from 'https://deno.land/std@0.168.0/testing/asserts.ts';
import {
  isApplePrivateRelayEmail,
  isSyntheticAuthEmail,
  isUsableContactEmail,
  publicContactEmailOrEmpty,
  resolveImportedContactEmail,
} from './contact_email.ts';

Deno.test('detecta os dois domínios Apple sem bloquear iCloud comum', () => {
  assertEquals(isApplePrivateRelayEmail('a@privaterelay.appleid.com'), true);
  assertEquals(isApplePrivateRelayEmail('a@private.icloud.com'), true);
  assertFalse(isApplePrivateRelayEmail('a@sub.private.icloud.com'));
  assertFalse(isApplePrivateRelayEmail('pessoa@icloud.com'));
  assertEquals(isUsableContactEmail('pessoa@icloud.com'), true);
});

Deno.test('detecta e-mail sintético do login por telefone', () => {
  assertEquals(isSyntheticAuthEmail('phone_5511987654321@stage.app'), true);
  assertFalse(isUsableContactEmail('phone_5511987654321@stage.app'));
  assertFalse(isSyntheticAuthEmail('pessoa@stage.app'));
});

Deno.test('normaliza contato público e suprime valores inseguros', () => {
  assertEquals(publicContactEmailOrEmpty('  Pessoa@Example.COM '), 'pessoa@example.com');
  assertEquals(publicContactEmailOrEmpty('alias@private.icloud.com'), '');
  assertEquals(publicContactEmailOrEmpty('inválido'), '');
});

Deno.test('importação preserva contato profissional existente', () => {
  assertEquals(
    resolveImportedContactEmail('manual@example.com', 'curriculo@example.com'),
    'manual@example.com',
  );
});

Deno.test('importação preenche vazio ou relay com contato válido', () => {
  assertEquals(
    resolveImportedContactEmail(null, 'CURRICULO@Example.com'),
    'curriculo@example.com',
  );
  assertEquals(
    resolveImportedContactEmail('alias@privaterelay.appleid.com', 'cv@example.com'),
    'cv@example.com',
  );
});

Deno.test('importação nunca introduz relay ou sintético', () => {
  assertEquals(resolveImportedContactEmail(null, 'alias@private.icloud.com'), null);
  assertEquals(resolveImportedContactEmail(null, 'phone_5511999999999@stage.app'), null);
});
